api/wft-complete-metadata.md

```diff
@@ -0,0 +1,126 @@
+    { command_type: "timer", id: 3, duration: "5s", location: "my_file.py:42" },
+]
+```
+
+### SDK internal versioning
+
+We would like to be able to make SDK changes which affect how history is written, which means
+flagging our behavior based on what version of the SDK the history was produced with. Note: *not*
+relating to concepts of versioning a *user's* code, but rather the SDK itself.
+
+This internal version could easily slot into this metadata.
+
+Note: This could potentially go in `temporal.api.taskqueue.v1.VersionId` too, but that's more
+intended to relate to user code.
+
+### Dynamic config adjustments
+
+Allow the worker to set configurable things which don't directly impact determinism or workflow
+state. For example, changing the WFT timeout for the next task.
+
+### Various debugging / analysis information
+
+There's a whole world of little bits of info that could be useful to enable selectively to help
+users debug their workflows, help us learn about performance information, etc.
+
+Some examples:
+* Emit the OTel span id that was associated with the WFT - now the user can go easily look up any
+  logs/events that were emitted as part of that WFT.
+* Metrics work great for all sorts of latency measurements, but what if you wanted to know about
+  how long a _specific_ WFT took? You could emit that.
+* Interceptors could pipe whatever arbitrary data the user wants into there, for use cases where
+  memo / search attributes / side effects may not be appropriate.
+
+### Things we haven't thought of
+
+Having this metadata be suitably generic means we dont need to update the API repo for every
+possible usecase. Particularly ones where the SDK is the only one who needs to be able to read it.
+
+## Implementation Proposal
+
+We'd add a new metadata field to the
+[WFT response](https://github.com/temporalio/api/blob/master/temporal/api/workflowservice/v1/request_response.proto#L274).
+And the [WFT completed event](https://github.com/temporalio/api/blob/f0350f8032ad2f0c60c539b3b61ea37f412f1cf7/temporal/api/history/v1/message.proto#L186).
+Though the completion event need not necessarily contain all the info sent in the completion. If
+it's known and understood that a particular metadata field is simply telling server something
+ephemeral, that could be stripped before writing the event.
+
+With the following type:
+
+```protobuf
+message RespondWorkflowTaskCompletedRequest {
+  // ...
+  
+  map<string, temporal.api.common.v1.Payload> metadata = 11;
+  
+  // ...
+}
+```
+
+
+### Why not just add more (specific) fields to WFT completion response?
+
+* Things that the SDK only needs to write down for itself (ex: the versioning usecase) can now be
+  made without requiring changes to the API repo, updating the server, releasing it, and ensuring
+  that that version is in use. That's real nice.
+* Supports arbitrary user-originated data (TBD if we actually want to allow this, but with this
```

Comment:
I think that simple arbitrary data associated with WFTs can be accomplished easily with tracing, and more advanced use cases should be left up to users. They can store their own data associated with a WFT id.